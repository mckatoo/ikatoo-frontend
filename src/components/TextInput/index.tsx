import { InputHTMLAttributes, useEffect, useState } from 'react'

import Styles from './styles'

export type TextInputProps = {
  onInputChange?: (value: string) => void
  label?: string
  labelColor?: 'black' | 'white'
  initialValue?: string
  icon?: React.ReactNode
  iconPosition?: 'left' | 'right'
  disabled?: boolean
  error?: string
  name: string
} & InputHTMLAttributes<HTMLInputElement>

const TextInput = ({
  icon,
  iconPosition = 'left',
  label,
  labelColor = 'black',
  name,
  initialValue = '',
  error,
  disabled = false,
  onInputChange,
  ...props
}: TextInputProps) => {
  const [value, setValue] = useState(initialValue)

  useEffect(() => {
    setValue(initialValue)
  }, [initialValue])

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = e.currentTarget.value
    setValue(newValue)

    !!onInputChange && onInputChange(newValue)
  }

  return (
    <Styles.Wrapper disabled={disabled} error={!!error}>
      {!!label && (
        <Styles.Label labelColor={labelColor} htmlFor={name}>
          {label}
        </Styles.Label>
      )}
      <Styles.InputWrapper>
        {!!icon && (
          <Styles.Icon iconPosition={iconPosition}>{icon}</Styles.Icon>
        )}
        <Styles.Input
          type="text"
          onChange={onChange}
          value={value}
          iconPosition={iconPosition}
          disabled={disabled}
          name={name}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </Styles.InputWrapper>
      {!!error && <Styles.Error>{error}</Styles.Error>}
    </Styles.Wrapper>
  )
}

export default TextInput
