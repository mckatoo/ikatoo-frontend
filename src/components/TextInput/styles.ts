import styled, { css } from 'styled-components'

import { TextInputProps } from '.'
import theme from '../../theme'

type IconPositionProps = Pick<TextInputProps, 'iconPosition'>

type LabelProps = Pick<TextInputProps, 'labelColor'>

type WrapperProps = Pick<TextInputProps, 'disabled'> & { error?: boolean }

const InputWrapper = styled.div`
  ${() => css`
    display: flex;
    align-items: center;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xsmall};
    border: 0.2rem solid;
    border-color: ${theme.colors.lightGray};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`

const Input = styled.input<IconPositionProps>`
  ${({ iconPosition }) => css`
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.large};
    padding: ${theme.spacings.xsmall} 0;
    padding-${iconPosition}: ${theme.spacings.xsmall};
    background: transparent;
    border: 0;
    outline: none;
    width: ${iconPosition === 'right' ? `calc(100% - 2.2rem)` : `100%`};
    &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0 ${theme.spacings.small}
        ${theme.colors.lightGray} inset;
      filter: none;
      &::first-line {
        font-family: ${theme.font.family};
        font-size: ${theme.font.sizes.large};
      }
    }
  `}
`

const Label = styled.label<LabelProps>`
  ${({ labelColor }) => css`
    font-size: ${theme.font.sizes.large};
    color: ${theme.colors[labelColor ?? 'black']};
    cursor: pointer;
  `}
`

const Icon = styled.div<IconPositionProps>`
  ${({ iconPosition }) => css`
    display: flex;
    color: ${theme.colors.gray};
    order: ${iconPosition === 'right' ? 1 : 0};

    & > svg {
      width: 1.2rem;
      height: 100%;
    }
  `}
`

const Error = styled.p`
  ${() => css`
    color: ${theme.colors.red};
    font-size: ${theme.font.sizes.xsmall};
  `}
`

const wrapperModifiers = {
  error: () => css`
    ${InputWrapper} {
      border-color: ${theme.colors.red};
    }

    ${Icon},
    ${Label} {
      color: ${theme.colors.red};
    }
  `,
  disabled: () => css`
    ${Label},
    ${Input},
    ${Icon} {
      cursor: not-allowed;
      color: ${theme.colors.gray};

      &::placeholder {
        color: currentColor;
      }
    }
  `
}

const Wrapper = styled.div<WrapperProps>`
  ${({ error, disabled }) => css`
    ${error && wrapperModifiers.error()}
    ${disabled && wrapperModifiers.disabled()}
  `}
`

export default {
  Wrapper,
  Error,
  Icon,
  Label,
  Input,
  InputWrapper
}
