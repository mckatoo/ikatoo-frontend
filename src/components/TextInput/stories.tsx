import { ComponentMeta, ComponentStory } from '@storybook/react'
import { Email } from '@styled-icons/material-outlined'
import TextInput from '.'

export default {
  title: 'Components/TextInput',
  component: TextInput,
  args: {
    label: 'E-mail',
    name: 'email',
    icon: <Email />,
    initialValue: '',
    placeholder: 'john.cage@gmail.com',
    disabled: false
  }
} as ComponentMeta<typeof TextInput>

export const Default: ComponentStory<typeof TextInput> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextInput {...args} />
  </div>
)

export const withError: ComponentStory<typeof TextInput> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextInput {...args} />
  </div>
)
withError.args = {
  error: 'Ops...something is wrong'
}
