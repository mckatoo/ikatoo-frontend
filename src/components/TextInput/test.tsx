import { render } from '@testing-library/react'
import { describe, expect, it } from 'vitest'
import TextInput from '.'

describe('TextInput', () => {
  it('should be able to render the text input', () => {
    const { getByRole } = render(<TextInput />)

    expect(getByRole('textbox')).toBeInTheDocument()
  })
})
