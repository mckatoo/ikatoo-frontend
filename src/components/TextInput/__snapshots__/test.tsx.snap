// Vitest Snapshot v1

exports[`<TextField /> > Renders with error 1`] = `
.c7 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #e5e7eb;
  border-radius: 0.2rem;
  padding: 0 0.4rem;
  border: 0.2rem solid;
  border-color: #e5e7eb;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #1e293b;
}

.c8 {
  color: #00021f;
  font-family: 'Oswald','Helvetica','Arial','sans-serif';
  font-size: 0.9rem;
  padding: 0.4rem 0;
  padding-left: 0.4rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c8:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 0.6rem #e5e7eb inset;
  -webkit-filter: none;
  filter: none;
}

.c8:-webkit-autofill::first-line {
  font-family: 'Oswald','Helvetica','Arial','sans-serif';
  font-size: 0.9rem;
}

.c2 {
  font-size: 0.9rem;
  color: #00021f;
  cursor: pointer;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #8F8F8F;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

.c6 > svg {
  width: 1.2rem;
  height: 100%;
}

.c9 {
  color: #ff2a2a;
  font-size: 0.6rem;
}

.c0 .c3 {
  border-color: #ff2a2a;
}

.c0 .c5,
.c0 .c1 {
  color: #ff2a2a;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="username"
  >
    TextField
  </label>
  <div
    class="c3 c4"
  >
    <div
      class="c5 c6"
    >
      <svg
        aria-hidden="true"
        class="c7"
        data-testid="icon"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M22 6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6zm-2 0-8 5-8-5h16zm0 12H4V8l8 5 8-5v10z"
        />
      </svg>
    </div>
    <input
      class="c8"
      id="username"
      name="username"
      type="text"
      value=""
    />
  </div>
  <p
    class="c9"
  >
    Error message
  </p>
</div>
`
